File: git-epic-creator/services/code_graph_ingestion_service/src/core/inventory.py
"""Deterministic file inventory builder (Task 05).

Produces stable file entries:
- path: repo-relative POSIX path
- sha256: sha256 over file bytes
- language: inferred from extension (conservative)
- line_count: physical line count based on bytes
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from pathlib import Path

from core.ignore_rules import IgnoreRules, build_ignore_rules


Language = str  # "cobol" | "java" | "javascript" | "xml" | "other"


@dataclass(frozen=True)
class InventoryEntry:
    path: str
    sha256: str
    language: Language
    line_count: int

def _count_physical_lines(data: bytes) -> int:
    if not data:
        return 0
    n = data.count(b"\n")
    if data.endswith(b"\n"):
        return n
    return n + 1


def _infer_language(rel_posix_path: str) -> Language:
    p = rel_posix_path.lower()
    if p.endswith((".cbl", ".cob", ".cpy")):
        return "cobol"
    if p.endswith(".java"):
        return "java"
    if p.endswith((".js", ".jsx", ".mjs", ".cjs", ".ts", ".tsx")):
        return "javascript"
    if p.endswith((".xml", ".xsd", ".wsdl")):
        return "xml"
    return "other"


def iter_repo_files(repo_root: Path, ignore: IgnoreRules) -> list[Path]:
    """Return a deterministic list of in-scope files under `repo_root`."""

    files: list[Path] = []
    for p in repo_root.rglob("*"):
        if not p.is_file():
            continue
        if p.is_symlink():
            continue
        rel = p.relative_to(repo_root).as_posix()
        if ignore.is_ignored(rel):
            continue
        files.append(p)
    return sorted(files, key=lambda x: x.relative_to(repo_root).as_posix())


def build_inventory(repo_root: Path, ignore: IgnoreRules | None = None) -> list[InventoryEntry]:
    """Build deterministic `files[]` inventory for `repo_index.json`."""

    ignore = ignore or build_ignore_rules(repo_root)
    out: list[InventoryEntry] = []

    for p in iter_repo_files(repo_root, ignore):
        rel = p.relative_to(repo_root).as_posix()
        data = p.read_bytes()
        out.append(
            InventoryEntry(
                path=rel,
                sha256=hashlib.sha256(data).hexdigest(),
                language=_infer_language(rel),
                line_count=_count_physical_lines(data),
            )
        )
    return out

File: git-epic-creator/services/code_graph_ingestion_service/src/core/inventory.rs
use crate::core::ignore_rules::{build_ignore_rules, IgnoreRules};
use crate::core::types::CodeLanguage;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::path::{Component, Path, PathBuf};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct InventoryEntry {
    pub path: String, // repo-relative POSIX path
    pub sha256: String,
    pub language: CodeLanguage,
    pub line_count: i64,
}

fn to_posix_rel_path(repo_root: &Path, p: &Path) -> anyhow::Result<String> {
    let rel = p.strip_prefix(repo_root)?;
    let mut out = String::new();
    for (i, c) in rel.components().enumerate() {
        let s = match c {
            Component::Normal(os) => os.to_string_lossy(),
            _ => continue,
        };
        if i > 0 {
            out.push('/');
        }
        out.push_str(&s);
    }
    Ok(out)
}

fn count_physical_lines(data: &[u8]) -> i64 {
    if data.is_empty() {
        return 0;
    }
    let n = data.iter().filter(|b| **b == b'\n').count() as i64;
    if data.last() == Some(&b'\n') {
        n
    } else {
        n + 1
    }
}

fn infer_language(rel_posix_path: &str) -> CodeLanguage {
    let p = rel_posix_path.to_ascii_lowercase();
    if p.ends_with(".cbl") || p.ends_with(".cob") || p.ends_with(".cpy") {
        return CodeLanguage::Cobol;
    }
    if p.ends_with(".java") {
        return CodeLanguage::Java;
    }
    if p.ends_with(".js")
        || p.ends_with(".jsx")
        || p.ends_with(".mjs")
        || p.ends_with(".cjs")
        || p.ends_with(".ts")
        || p.ends_with(".tsx")
    {
        return CodeLanguage::Javascript;
    }
    if p.ends_with(".xml") || p.ends_with(".xsd") || p.ends_with(".wsdl") {
        return CodeLanguage::Xml;
    }
    CodeLanguage::Other
}

pub fn iter_repo_files(repo_root: &Path, ignore: &IgnoreRules) -> anyhow::Result<Vec<PathBuf>> {
    let mut files: Vec<PathBuf> = Vec::new();
    for entry in walkdir::WalkDir::new(repo_root).follow_links(false) {
        let entry = entry?;
        if !entry.file_type().is_file() {
            continue;
        }
        let p = entry.path().to_path_buf();
        let rel = to_posix_rel_path(repo_root, &p)?;
        if ignore.is_ignored(&rel) {
            continue;
        }
        files.push(p);
    }
    files.sort_by_key(|p| to_posix_rel_path(repo_root, p).unwrap_or_default());
    Ok(files)
}

pub fn build_inventory(repo_root: &Path, ignore: Option<&IgnoreRules>) -> anyhow::Result<Vec<InventoryEntry>> {
    let owned_ignore;
    let ignore = match ignore {
        Some(i) => i,
        None => {
            owned_ignore = build_ignore_rules(repo_root, None)?;
            &owned_ignore
        }
    };

    let mut out: Vec<InventoryEntry> = Vec::new();
    for p in iter_repo_files(repo_root, ignore)? {
        let rel = to_posix_rel_path(repo_root, &p)?;
        let data = std::fs::read(&p)?;
        let sha = hex::encode(Sha256::digest(&data));
        out.push(InventoryEntry {
            path: rel.clone(),
            sha256: sha,
            language: infer_language(&rel),
            line_count: count_physical_lines(&data),
        });
    }
    Ok(out)
}


